File: benchmarks/c/string_utils.c
/*
 * String Utilities - Benchmark C Program
 * String manipulation functions for test generation evaluation
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

// Get string length (safe version)
size_t safe_strlen(const char* str) {
    if (str == NULL) return 0;
    return strlen(str);
}

// Compare strings (case insensitive)
int strcasecmp_safe(const char* s1, const char* s2) {
    if (s1 == NULL && s2 == NULL) return 0;
    if (s1 == NULL) return -1;
    if (s2 == NULL) return 1;

    while (*s1 && *s2) {
        int c1 = tolower((unsigned char)*s1);
        int c2 = tolower((unsigned char)*s2);
        if (c1 != c2) return c1 - c2;
        s1++;
        s2++;
    }
    return tolower((unsigned char)*s1) - tolower((unsigned char)*s2);
}

// Check if string starts with prefix
int starts_with(const char* str, const char* prefix) {
    if (str == NULL || prefix == NULL) return 0;

    while (*prefix) {
        if (*str != *prefix) return 0;
        str++;
        prefix++;
    }
    return 1;
}

// Check if string ends with suffix
int ends_with(const char* str, const char* suffix) {
    if (str == NULL || suffix == NULL) return 0;

    size_t str_len = strlen(str);
    size_t suffix_len = strlen(suffix);

    if (suffix_len > str_len) return 0;

    return strcmp(str + str_len - suffix_len, suffix) == 0;
}

// Count occurrences of character in string
int count_char(const char* str, char c) {
    if (str == NULL) return 0;

    int count = 0;
    while (*str) {
        if (*str == c) count++;
        str++;
    }
    return count;
}

// Count occurrences of substring
int count_substring(const char* str, const char* sub) {
    if (str == NULL || sub == NULL || *sub == '\0') return 0;

    int count = 0;
    size_t sub_len = strlen(sub);
    const char* pos = str;

    while ((pos = strstr(pos, sub)) != NULL) {
        count++;
        pos += sub_len;
    }
    return count;
}

// Reverse string in place
void reverse_string(char* str) {
    if (str == NULL) return;

    size_t len = strlen(str);
    if (len <= 1) return;

    char* start = str;
    char* end = str + len - 1;

    while (start < end) {
        char temp = *start;
        *start = *end;
        *end = temp;
        start++;
        end--;
    }
}

// Convert string to uppercase (in place)
void to_uppercase(char* str) {
    if (str == NULL) return;

    while (*str) {
        *str = toupper((unsigned char)*str);
        str++;
    }
}

// Convert string to lowercase (in place)
void to_lowercase(char* str) {
    if (str == NULL) return;

    while (*str) {
        *str = tolower((unsigned char)*str);
        str++;
    }
}

// Trim leading whitespace (returns pointer within string)
char* trim_left(char* str) {
    if (str == NULL) return NULL;

    while (isspace((unsigned char)*str)) {
        str++;
    }
    return str;
}

// Trim trailing whitespace (modifies string)
void trim_right(char* str) {
    if (str == NULL) return;

    size_t len = strlen(str);
    while (len > 0 && isspace((unsigned char)str[len - 1])) {
        str[--len] = '\0';
    }
}

// Check if string is palindrome
int is_palindrome(const char* str) {
    if (str == NULL) return 0;

    size_t len = strlen(str);
    if (len <= 1) return 1;

    const char* left = str;
    const char* right = str + len - 1;

    while (left < right) {
        if (*left != *right) return 0;
        left++;
        right--;
    }
    return 1;
}

// Check if string contains only digits
int is_numeric(const char* str) {
    if (str == NULL || *str == '\0') return 0;

    // Allow leading minus sign
    if (*str == '-' || *str == '+') {
        str++;
        if (*str == '\0') return 0;
    }

    while (*str) {
        if (!isdigit((unsigned char)*str)) return 0;
        str++;
    }
    return 1;
}

// Check if string contains only alphabetic characters
int is_alpha(const char* str) {
    if (str == NULL || *str == '\0') return 0;

    while (*str) {
        if (!isalpha((unsigned char)*str)) return 0;
        str++;
    }
    return 1;
}

// Check if string contains only alphanumeric characters
int is_alnum(const char* str) {
    if (str == NULL || *str == '\0') return 0;

    while (*str) {
        if (!isalnum((unsigned char)*str)) return 0;
        str++;
    }
    return 1;
}

// Find first occurrence of any character from set
int find_first_of(const char* str, const char* chars) {
    if (str == NULL || chars == NULL) return -1;

    const char* p = str;
    while (*p) {
        const char* c = chars;
        while (*c) {
            if (*p == *c) return (int)(p - str);
            c++;
        }
        p++;
    }
    return -1;
}

// Replace first occurrence of old_char with new_char
int replace_char(char* str, char old_char, char new_char) {
    if (str == NULL) return 0;

    while (*str) {
        if (*str == old_char) {
            *str = new_char;
            return 1;  // Replaced
        }
        str++;
    }
    return 0;  // Not found
}

// Replace all occurrences of old_char with new_char
int replace_all_char(char* str, char old_char, char new_char) {
    if (str == NULL) return 0;

    int count = 0;
    while (*str) {
        if (*str == old_char) {
            *str = new_char;
            count++;
        }
        str++;
    }
    return count;
}

File: benchmarks/rust/string_utils.rs
//! String Utilities - Benchmark Rust Program
//! String manipulation functions for test generation evaluation

/// Check if string starts with prefix
pub fn starts_with(s: &str, prefix: &str) -> bool {
    s.starts_with(prefix)
}

/// Check if string ends with suffix
pub fn ends_with(s: &str, suffix: &str) -> bool {
    s.ends_with(suffix)
}

/// Count occurrences of character in string
pub fn count_char(s: &str, c: char) -> usize {
    s.chars().filter(|&ch| ch == c).count()
}

/// Count occurrences of substring
pub fn count_substring(s: &str, sub: &str) -> usize {
    if sub.is_empty() {
        return 0;
    }
    s.matches(sub).count()
}

/// Reverse a string
pub fn reverse_string(s: &str) -> String {
    s.chars().rev().collect()
}

/// Convert string to uppercase
pub fn to_uppercase(s: &str) -> String {
    s.to_uppercase()
}

/// Convert string to lowercase
pub fn to_lowercase(s: &str) -> String {
    s.to_lowercase()
}

/// Trim whitespace from both ends
pub fn trim(s: &str) -> &str {
    s.trim()
}

/// Check if string is palindrome
pub fn is_palindrome(s: &str) -> bool {
    let chars: Vec<char> = s.chars().collect();
    let len = chars.len();

    if len <= 1 {
        return true;
    }

    for i in 0..len / 2 {
        if chars[i] != chars[len - 1 - i] {
            return false;
        }
    }
    true
}

/// Check if string is palindrome (case insensitive, alphanumeric only)
pub fn is_palindrome_normalized(s: &str) -> bool {
    let chars: Vec<char> = s
        .chars()
        .filter(|c| c.is_alphanumeric())
        .map(|c| c.to_ascii_lowercase())
        .collect();

    let len = chars.len();
    if len <= 1 {
        return true;
    }

    for i in 0..len / 2 {
        if chars[i] != chars[len - 1 - i] {
            return false;
        }
    }
    true
}

/// Check if string contains only digits
pub fn is_numeric(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }

    let mut chars = s.chars().peekable();

    // Allow leading sign
    if let Some(&c) = chars.peek() {
        if c == '-' || c == '+' {
            chars.next();
            if chars.peek().is_none() {
                return false;
            }
        }
    }

    chars.all(|c| c.is_ascii_digit())
}

/// Check if string contains only alphabetic characters
pub fn is_alpha(s: &str) -> bool {
    !s.is_empty() && s.chars().all(|c| c.is_alphabetic())
}

/// Check if string contains only alphanumeric characters
pub fn is_alnum(s: &str) -> bool {
    !s.is_empty() && s.chars().all(|c| c.is_alphanumeric())
}

/// Find first occurrence of any character from set
pub fn find_first_of(s: &str, chars: &str) -> Option<usize> {
    s.find(|c| chars.contains(c))
}

/// Replace first occurrence of pattern
pub fn replace_first(s: &str, from: &str, to: &str) -> String {
    if from.is_empty() {
        return s.to_string();
    }

    match s.find(from) {
        Some(pos) => {
            let mut result = String::with_capacity(s.len() - from.len() + to.len());
            result.push_str(&s[..pos]);
            result.push_str(to);
            result.push_str(&s[pos + from.len()..]);
            result
        }
        None => s.to_string(),
    }
}

/// Replace all occurrences of pattern
pub fn replace_all(s: &str, from: &str, to: &str) -> String {
    if from.is_empty() {
        return s.to_string();
    }
    s.replace(from, to)
}

/// Split string into words (whitespace separated)
pub fn split_words(s: &str) -> Vec<&str> {
    s.split_whitespace().collect()
}

/// Count words in string
pub fn word_count(s: &str) -> usize {
    s.split_whitespace().count()
}

/// Capitalize first letter of string
pub fn capitalize(s: &str) -> String {
    let mut chars = s.chars();
    match chars.next() {
        None => String::new(),
        Some(first) => first.to_uppercase().chain(chars).collect(),
    }
}

/// Capitalize first letter of each word
pub fn title_case(s: &str) -> String {
    s.split_whitespace()
        .map(|word| capitalize(word))
        .collect::<Vec<_>>()
        .join(" ")
}

/// Check if string is empty or whitespace only
pub fn is_blank(s: &str) -> bool {
    s.trim().is_empty()
}

/// Truncate string to max length, adding ellipsis if needed
pub fn truncate(s: &str, max_len: usize) -> String {
    if s.len() <= max_len {
        s.to_string()
    } else if max_len <= 3 {
        s.chars().take(max_len).collect()
    } else {
        let mut result: String = s.chars().take(max_len - 3).collect();
        result.push_str("...");
        result
    }
}

/// Repeat string n times
pub fn repeat(s: &str, n: usize) -> String {
    s.repeat(n)
}

/// Check if two strings are anagrams
pub fn are_anagrams(s1: &str, s2: &str) -> bool {
    if s1.len() != s2.len() {
        return false;
    }

    let mut chars1: Vec<char> = s1.chars().collect();
    let mut chars2: Vec<char> = s2.chars().collect();

    chars1.sort_unstable();
    chars2.sort_unstable();

    chars1 == chars2
}

/// Levenshtein distance between two strings
pub fn levenshtein_distance(s1: &str, s2: &str) -> usize {
    let s1_chars: Vec<char> = s1.chars().collect();
    let s2_chars: Vec<char> = s2.chars().collect();

    let m = s1_chars.len();
    let n = s2_chars.len();

    if m == 0 {
        return n;
    }
    if n == 0 {
        return m;
    }

    let mut prev: Vec<usize> = (0..=n).collect();
    let mut curr = vec![0; n + 1];

    for i in 1..=m {
        curr[0] = i;
        for j in 1..=n {
            let cost = if s1_chars[i - 1] == s2_chars[j - 1] {
                0
            } else {
                1
            };
            curr[j] = (prev[j] + 1)
                .min(curr[j - 1] + 1)
                .min(prev[j - 1] + cost);
        }
        std::mem::swap(&mut prev, &mut curr);
    }

    prev[n]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_reverse_string() {
        assert_eq!(reverse_string("hello"), "olleh");
        assert_eq!(reverse_string(""), "");
    }
}
